packages/jest-config/src/readConfigFileAndSetRootDir.ts

```diff
@@ -10,26 +10,55 @@ import * as fs from 'fs';
 import {Config} from '@jest/types';
 // @ts-ignore: vendored
 import jsonlint from './vendor/jsonlint';
-import {PACKAGE_JSON} from './constants';
+import {
+  JEST_CONFIG_EXT_JSON,
+  JEST_CONFIG_EXT_MJS,
+  PACKAGE_JSON,
+} from './constants';
 
 // Read the configuration and set its `rootDir`
 // 1. If it's a `package.json` file, we look into its "jest" property
 // 2. For any other file, we just require it.
-export default (configPath: Config.Path): Config.InitialOptions => {
-  const isJSON = configPath.endsWith('.json');
+export default async (
+  configPath: Config.Path,
+): Promise<Config.InitialOptions> => {
+  const isJSON = configPath.endsWith(JEST_CONFIG_EXT_JSON);
+  const isMjs = configPath.endsWith(JEST_CONFIG_EXT_MJS);
   let configObject;
 
-  try {
-    configObject = require(configPath);
-  } catch (error) {
-    if (isJSON) {
-      throw new Error(
-        `Jest: Failed to parse config file ${configPath}\n` +
-          `  ${jsonlint.errors(fs.readFileSync(configPath, 'utf8'))}`,
-      );
-    } else {
+  if (isMjs) {
+    try {
+      const importedConfig = await import(configPath);
+
+      if (!importedConfig.default) {
+        throw new Error(
+          `Jest: Failed to load mjs config file ${configPath} - did you use a default export?`,
+        );
+      }
+
+      configObject = importedConfig.default;
+    } catch (error) {
+      if (error.message === 'Not supported') {
```

Comment:
afaik there's no clean way of detecting support for `import` function before trying to use it. I think this is fine, though